content/recipes/authoring/kata-snippets/full-tests.md

```diff
@@ -7,4 +7,57 @@ next: /recipes/authoring/kata-snippets/preloaded/
+- **Tests should not be susceptible to input mutation.** It should not be possible for user solution to modify test input in a way which would influence execution and result of the test. If reference solution is used, it should be called first, and user solution should be called afterwards with the same input.
+- **Reference solution should not be revealed to user.** When assertion fails or test crashes, some test frameworks print fragment of source code which caused failure to the console. It may happen that such printed failure message or stack trace exposes information about solution which should not be revealed.
+- **Reference solution shouldn't be accessible to user solution.** It should not be possible to implement user solution as an alias or wrapper around reference solution used by tests. Reference solution should not be possible to be called by any other part of the kata other than the tests. In particular, it should not be a global function.
+- **Test suite should not take long time to run.** For kata which are not performance oriented, size of the test suite should allow it to be completed in just a few (ideally 3-4) seconds. There's a time limit for every run (for some language  it is 12 seconds, for others it's 16 seconds), but generation of test cases, invovations of user solution and calls to reference solution (when used) should run well below this limit. Otherwise there's a risk that tests will inconsistently time out, or some valid, but not optimally performant solutions will be rejected.
+
+## Fixed Tests
+
+Fixed tests are tests with predetermined inputs, and do not change between test runs.
+
+- **Each requirement should have corresponding fixed test.** Every requirement of the kata described in the description should be explicitly tested by at least one (possibly more) dedicated, properly labeled test case(s).
+- If your test suite uses random tests (and it usually should, see below), then make sure that every test scenario generated there has its equivalent fixed test. When random test fails, it should be accompanied by at least one failed equivalent fixed test.
+
+## Random Tests
+
+Random tests are not common testing practice and are somewhat specific to Codewars. They are required by many kata to reject solutions based on input probing, hardcoding, and other workarounds. Goal of random tests is to make input of test cases unpredictable, so only solutions which actually solve the problem can pass. Random tests are necessary especialy for difficult, highly rewarding kata, but Codewars testing guidelines require them for every kata they are applicable for.
+
+- **Each requirement should have at least one (preferrably more) corresponding random test case generated.** Basically, randomly generated scenarios should correspond to the set of fixed tests.
```

Comment:
```suggestion

### Random tests: definition

Tests are considered random when they match the following behaviours:
* When the test suite is run one time, the inputs of the test are unpredictable.
* When the test suite is run one time, the outputs of the test are equally unpredictable.
* When the tests are run several times, the order of the inputs/outputs is different, and inputs/outputs showing up in one run may not appear in other runs.

Hence, different batches of random tests where one expects one kind of behavior (test only primes, or successful tests), then another expect another behaiour (test only composite numbers or expecting errors) aren't random tests.

### Randomized tests: definition

In some rare situations, it may be  either not possible or too complex to actually build a satistfaying random generator. Or even, the domain of the testable inputs is too small. In those specific **but rare** cases, randomized tests are acceptable. Tests are considered randomized when:
* The inputs (possibly the output) are hardcoded
* From on run to the other, only the order of the inputs/outputs vary

Keep in mind that randomized tests can easily be hardcoded so are to be avoided as much as possible, to the benefit of real random tests. One way to hide a bit the randomized testing could be to have a set of `N` inputs, and build the randomized tests by picking randomly only a fraction of those inputs. But again, true random tests are better.

### Guidelines for random testing

- **Each requirement should have at least one (preferrably more) corresponding random test case generated.** Basically, randomly generated scenarios should correspond to the set of fixed tests.
```